File: py/operators.py

import math

epsilon = 1e-6

# core mathematical operators

def mul(x: float, y: float) -> float:
    return x * y

assert mul(2, 3) == 6

def id(x: float) -> float:
    return x

assert id(3) == 3

def add(x: float, y: float) -> float:
    return x + y

assert add(2, 3) == 5

def neg(x: float) -> float:
    return -x

assert neg(3) == -3

def lt(x: float, y: float) -> float:
    return 1.0 if x < y else 0.0

assert lt(2, 3) == 1.0

def eq(x: float, y: float) -> float:
    return 1.0 if x == y else 0.0

assert eq(2, 3) == 0.0

def max(x: float, y: float) -> float:
    return x if x > y else y

assert max(2, 3) == 3

def is_close(x: float, y: float) -> float:
    return abs(x - y) < 1e-2

assert is_close(2, 3) == 0.0

def sigmoid(x: float) -> float:
    return 1.0 / (1.0 + math.exp(-x)) if x >= 0 else math.exp(x) / (1.0 + math.exp(x))

assert is_close(sigmoid(2), 0.88)

def relu(x: float) -> float:
    return x if x > 0 else 0.0

assert relu(2) == 2

def log(x: float) -> float:
    return math.log(x + epsilon)

assert is_close(log(2), 0.69)

def exp(x: float) -> float:
    return math.exp(x)

assert is_close(exp(2), 7.39)

def log_back(x: float, d: float) -> float:
    return d / (x + epsilon)

assert is_close(log_back(2, 1), 0.5)

def inv(x: float) -> float:
    return 1.0 / x

assert is_close(inv(2), 0.5)

def inv_back(x: float, d: float) -> float:
    return -d / (x * x)

assert is_close(inv_back(2, 1), -0.25)

def relu_back(x: float, d: float) -> float:
    return d if x > 0 else 0.0

assert relu_back(2, 1) == 1

def sigmoid_back(x: float, d: float) -> float:
    return d * math.exp(-x) / ((1 + math.exp(-x)) ** 2)

assert is_close(sigmoid_back(2, 1), 0.11)

# higher-order functions

def map(fn):
    return lambda ls: [fn(x) for x in ls]

assert map(neg)([1, 2, 3]) == [-1, -2, -3]

def neg_list(ls):
    return map(neg)(ls)

assert neg_list([1, 2, 3]) == [-1, -2, -3]

def map2(fn):
    return lambda ls1, ls2: [fn(x, y) for x, y in zip(ls1, ls2)]

assert map2(add)([1, 2, 3], [4, 5, 6]) == [5, 7, 9]

def add_list(ls1, ls2):
    return map2(add)(ls1, ls2)

assert add_list([1, 2, 3], [4, 5, 6]) == [5, 7, 9]

def reduce(fn, start):
    def reduce_fn(ls, fn, start):
        iterator = iter(ls)
        for i in iterator:
            start = fn(start, i)
        return start

    return lambda ls: reduce_fn(ls, fn, start)

assert reduce(mul, 2)([1, 2, 3]) == 12

def sum(ls):
    return reduce(add, 0)(ls)

assert sum([1, 2, 3]) == 6

def prod(ls):
    return reduce(mul, 1)(ls)

assert prod([2, 3, 4]) == 24

File: rb/operators.rb

$epsilon = 1e-6

# core mathematical operators
    
def mul(x, y)
    return x.to_f * y
end

raise Exception unless mul(2, 3) == 6

def id(x)
    return x.to_f
end

raise Exception unless id(3) == 3

def add(x, y)
    return x.to_f + y
end

raise Exception unless add(2, 3) == 5

def neg(x)
    return -x.to_f
end

raise Exception unless neg(3) == -3

def lt(x, y)
    return x < y ? 1.0 : 0.0
end

raise Exception unless lt(2, 3) == 1

def eq(x, y)
    return x == y ? 1.0 : 0.0
end

raise Exception unless eq(2, 3) == 0

def max(x, y)
    return x > y ? x.to_f : y.to_f
end

raise Exception unless max(2, 3) == 3

def is_close(x, y)
    return (x - y).abs.to_f < 1e-2
end

raise Exception unless !is_close(2, 3)

def sigmoid(x)
    return x >= 0 ? 1.0 / (1.0 + Math.exp(-x)) : Math.exp(x) / (1.0 + Math.exp(x)).to_f
end

raise Exception unless is_close(sigmoid(2), 0.88)

def relu(x)
    return x >= 0 ? x.to_f : 0.0
end

raise Exception unless relu(2) == 2

def log(x)
    return Math.log(x + $epsilon).to_f
end

raise Exception unless is_close(log(2), 0.69)

def exp(x)
    return Math.exp(x).to_f
end

raise Exception unless is_close(exp(2), 7.39)

def log_back(x, d)
    return d.to_f / (x + $epsilon)
end

raise Exception unless is_close(log_back(2, 1), 0.5)

def inv(x)
    return 1.0 / x.to_f
end

raise Exception unless is_close(inv(2), 0.5)

def inv_back(x, d)
    return -d / (x * x).to_f
end

raise Exception unless is_close(inv_back(2, 1), -0.25)

def relu_back(x, d)
    return x > 0 ? d.to_f : 0.0
end

raise Exception unless relu_back(2, 1) == 1

def sigmoid_back(x, d)
    return d.to_f * Math.exp(-x) / ((1.0 + Math.exp(-x)) ** 2)
end

raise Exception unless is_close(sigmoid_back(2, 1), 0.11)

# higher-order functions

def map(fn, ls)
    return ls.map { |x| fn.call(x) }
end

raise Exception unless map(method(:neg), [1, 2, 3]) == [-1, -2, -3]

def neg_list(ls)
    return map(method(:neg), ls)
end

raise Exception unless neg_list([1, 2, 3]) == [-1, -2, -3]

def map2(fn, ls1, ls2)
    return ls1.zip(ls2).map { |x, y| fn.call(x, y) }
end

raise Exception unless map2(method(:add), [1, 2, 3], [4, 5, 6]) == [5, 7, 9]

def add_list(ls1, ls2)
    return map2(method(:add), ls1, ls2)
end

raise Exception unless add_list([1, 2, 3], [4, 5, 6]) == [5, 7, 9]

def reduce(fn, start, ls)
    def reduce_fn(ls, fn, start)
        for i in 0...ls.length
            start = fn.call(start, ls[i])
        end
        return start
    end
    return reduce_fn(ls, fn, start)
end

raise Exception unless reduce(method(:mul), 2, [1, 2, 3]) == 12

def sum(ls)
    return reduce(method(:add), 0, ls)
end

raise Exception unless sum([1, 2, 3]) == 6

def prod(ls)
    return reduce(method(:mul), 1, ls)
end

raise Exception unless prod([2, 3, 4]) == 24
